contracts/Marketplace2.sol

```diff
@@ -0,0 +1,420 @@
+        _initialize(datacoinAddress, currencyUpdateAgentAddress, prev_marketplace_address);
+    }
+
+    function _initialize(address datacoinAddress, address currencyUpdateAgentAddress, address prev_marketplace_address) internal {
+        currencyUpdateAgent = currencyUpdateAgentAddress;
+        datacoin = ERC20(datacoinAddress);
+        prev_marketplace = IMarketplace(prev_marketplace_address);
+    }
+
+    ////////////////// Product management /////////////////
+
+    mapping (bytes32 => Product) public products;
+    /*
+        checks this marketplace first, then the previous
+    */
+    function getProduct(bytes32 id) public view returns (string name, address owner, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds, ProductState state) {
+        (name, owner, beneficiary, pricePerSecond, currency, minimumSubscriptionSeconds, state) = _getProductLocal(id);
+        if (owner != 0x0)
+            return;
+        return prev_marketplace.getProduct(id);
+    }
+
+    /**
+    checks only this marketplace, not the previous marketplace
+     */
+
+    function _getProductLocal(bytes32 id) internal view returns (string name, address owner, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds, ProductState state) {
+
+        return (
+            products[id].name,
+            products[id].owner,
+            products[id].beneficiary,
+            products[id].pricePerSecond,
+            products[id].priceCurrency,
+            products[id].minimumSubscriptionSeconds,
+            products[id].state
+        );
+    }
+
+    // also checks that p exists: p.owner == 0 for non-existent products
+    modifier onlyProductOwner(bytes32 productId) {
+        (,address _owner,,,,,) = getProduct(productId);
+        require(_owner != 0x0, "error_notFound");
+        require(_owner == msg.sender || owner == msg.sender, "error_productOwnersOnly");
+        _;
+    }
+
+    /**
+     * Imports product details (but NOT subscription details) from previous marketplace
+     */
+    function _importProductIfNeeded(bytes32 productId) internal returns (bool imported){
+        Product storage p = products[productId];
+        if(p.id != 0x0)
+            return false;
+        (string memory _name, address _owner, address _beneficiary, uint _pricePerSecond, IMarketplace.Currency _priceCurrency, uint _minimumSubscriptionSeconds, IMarketplace.ProductState _state) = prev_marketplace.getProduct(productId);
+        if(_owner == 0x0)
+            return false;
+        p.id = productId;
+        p.name = _name;
+        p.owner = _owner;
+        p.beneficiary = _beneficiary;
+        p.pricePerSecond = _pricePerSecond;
+        p.priceCurrency = _priceCurrency;
+        p.minimumSubscriptionSeconds = _minimumSubscriptionSeconds;
+        p.state = _state;
+//        emit ProductCreated(msg.sender, id, name, beneficiary, pricePerSecond, currency, minimumSubscriptionSeconds);
```

Comment:
out-commented lines should be cleaned up

Reply:
done